services/search/tests/routes/test_filter.py

```diff
@@ -0,0 +1,94 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2023 The HuggingFace Authors.
+
+import os
+from collections.abc import Generator
+
+import duckdb
+import pyarrow as pa
+import pytest
+from datasets import Dataset
+
+from search.routes.filter import execute_filter_query
+
+
+@pytest.fixture
+def ds() -> Dataset:
+    return Dataset.from_dict(
+        {
+            "name": ["Marie", "Paul", "Leo", "Simone"],
+            "gender": ["female", "male", "male", "female"],
+            "age": [35, 30, 25, 30],
+        }
+    )
+
+
+@pytest.fixture  # TODO: session scope
+def index_file_location(ds: Dataset) -> Generator[str, None, None]:
+    index_file_location = "index.duckdb"
+    con = duckdb.connect(index_file_location)
+    con.execute("INSTALL 'httpfs';")
+    con.execute("LOAD 'httpfs';")
+    con.execute("INSTALL 'fts';")
+    con.execute("LOAD 'fts';")
+    con.sql("CREATE OR REPLACE SEQUENCE serial START 0 MINVALUE 0;")
+    sample_df = ds.to_pandas()  # noqa: F841
+    create_command_sql = "CREATE OR REPLACE TABLE data AS SELECT nextval('serial') AS __hf_index_id, * FROM sample_df"
+    con.sql(create_command_sql)
+    # assert sample_df.shape[0] == con.execute(query="SELECT COUNT(*) FROM data;").fetchall()[0][0]
+    con.sql("PRAGMA create_fts_index('data', '__hf_index_id', 'name', 'gender', overwrite=1);")
+    con.close()
+    yield index_file_location
+    os.remove(index_file_location)
+
+
+def test_execute_filter_query(index_file_location: str) -> None:
+    columns, where, limit, offset = ["name", "age"], "gender = 'female'", 1, 1
+    num_rows_total, pa_table = execute_filter_query(
+        index_file_location=index_file_location, columns=columns, where=where, limit=limit, offset=offset
+    )
+    assert num_rows_total == 2
+    assert pa_table == pa.Table.from_pydict({"name": ["Simone"], "age": [30]})
```

Comment:
related to the comment above, on just passing the `where` parameter to duckdb. I think we want:
- to do a lot of tests, at least for every supported column type, and for several operators
- to pass wrong or "dangerous" queries and see what happens

Reply:
See `test_validate_where_parameter`, `test_validate_where_parameter_raises`, `test_execute_filter_query` and `test_execute_filter_query_raises` in `test_filter.py` file.

Reply:
OK. Maybe add more tests in the next PRs if we get bugs (test to reproduce the bug, and check it's fixed)